### DISCUSSION TITLE
Pyright playground?

### DISCUSSION THREAD

**Discussion by decorator-factory** [microsoft/pyright#2624] [Ideas]
I'd like to make a website similar to https://mypy-play.net/ but for Pyright.

Would it be OK for me to claim a domain name such as `pyright-play.net` that contains the word "pyright" in it?

**Comment by erictraut** [comment]
Thanks for the inquiry. I'll discuss with my colleagues and get back to you on Monday.

**Reply by erictraut** [reply]
We were thinking of creating a pyright playground but haven't gotten around to it. If you want to build one, that would be great!

You're welcome to claim a domain name that includes `pyright`. We didn't trademark that name, so there are no legal concerns. Our only ask is that you provide a statement somewhere on the site (e.g., in the README) that indicates that it isn't run by Microsoft or the maintainers of pyright.

Let us know if you run into any problems getting it set up. I look forward to seeing it working!

**Reply by decorator-factory** [reply]
Thank you Eric! I think I'll hack together something in the following week or two

**Reply by erictraut** [reply]
Are you planning to use the [monaco editor](https://microsoft.github.io/monaco-editor/)? It has all the hooks for language servers, so features like hover text could be easily added in the playground, which could be really cool. If you're interested in an example of a playground based on monaco, you can check out the [typescript playground](https://www.typescriptlang.org/play). If you need help integrating monaco, let me know because I have experience doing this. 

It looks like mypy's playground uses a different web-based editor. I don't know if it has the same capabilities as monaco.

Another design consideration is whether you plan to run pyright in the browser or on the server. I'm guessing that the plan is to run it on the server so it's running in a node environment. That has certain advantages (e.g. you can install an arbitrary version of the pyright CLI from npm and you don't need to worry about file system accesses for type stubs), but it also means that the playground won't be as interactive (because of the round trip to the server). It's possible you could get pyright to run as a language server within the browser, at least for recent versions of pyright. That's because work has been done over the past several months to reduce and even eliminate the node dependencies so pyright (or rather pylance) can run in vscode.dev, the web-based version of vscode. It might be best to start with a server-based design because running it in the browser adds quite a bit of complexity to the implementation.

In any case, let us know what design decisions you are making, and we can assist accordingly. Or if you want to discuss the pros/cons of certain design decisions, I'm happy to have those discussions.

**Reply by decorator-factory** [reply]
I was originally planning to make it similar to mypy-play.net (except faster, probably using watch mode on the server?) so I'd use the latest CodeMirror editor. I think that's what I'll do first. But monaco & the language server in the browser sounds like a good upgrade.

Do you think it's feasible to take a mixed approach? For example, run a language server on the server and forward all the responses for a given file via websockets. That would allow switching between versions and installing popular libraries (or their stubs), while keeping the frontend more lightweight.

**Reply by erictraut** [reply]
Yeah, a mixed approach sounds good to me. It won't be as responsive as if it is running locally, but it should be good enough for most users who have low-latency internet connections.

**Reply by decorator-factory** [reply]
Monaco also will not work on mobile, it seems. Although that seems like a niche use case.

**Reply by erictraut** [reply]
I've used both CodeMirror and Monaco on various projects. Monaco is _much_ better IMO. If you're choosing between those two, I highly recommend choosing Monaco. It has many useful features that CodeMirror lacks.

**Reply by decorator-factory** [reply]
So this is what I came up with:
https://pyright-playground.decorator-factory.su/ (haven't registered a new domain name yet, so I'm using a subdomain :) )
It doesn't have any configuration or the save-to-gist-feature, but seems like the core is working
![demonstration](https://raw.githubusercontent.com/decorator-factory/pyright-playground/master/demo.gif)

How it works:
1. When the code changes, the client sends a POST request to the server with the code
2. The requests are debounced so that they're made at most once per second
3. The server then returns the JSON that the pyright CLI produced.

For development, the backed from the repo can be used ([decorator-factory/pyright-playground](https://github.com/decorator-factory/pyright-playground)). But for production i'm proxying to a serverless function.

**Reply by erictraut** [reply]
That's great progress! It's looking really good.

Have you thought about supporting any form of persistence so people can share their code snippets?

**Reply by decorator-factory** [reply]
re editor choice: I tried both Monaco and CodeMirror.
The Monaco editor seems really heavy, it takes a couple of minutes to build on my VPS, and about 15 seconds while hot-reloading on my personal computer. I also couldn't figure out how to make it responsive.

Maybe Monaco will turn out to be more useful if we use Pyright's language server feature. But for now I switched to CodeMirror

**Reply by decorator-factory** [reply]
> Have you thought about supporting any form of persistence so people can share their code snippets?

Yeah, I have that on the TODO list. I think i'll do it via private cloud storage and not a gist. Or maybe I'll add two options:
- Run a user-created gist
- Save to/load from the cloud using a short name

**Reply by decorator-factory** [reply]
It's a bit of a hack, but now you can use a source:
- a github issue (picks up the first Python code block): [?issue=microsoft/pyright/2662](https://pyright-playground.decorator-factory.su/?issue=microsoft/pyright/2662)
- a gist: [?gist_id=b00b00716dbc9b2115693a1ac8c7db54&filename=example.py](https://pyright-playground.decorator-factory.su/?gist_id=b00b00716dbc9b2115693a1ac8c7db54&filename=example.py)
- a base64+gzip string: [?gzip=H4sIAPHOr2EC_zXHPQqAMAwG0D2n-OjUgl6goKcI7oKNdOgPMUJ7eyff9kRbgc2e641celMDz56OU4kY2x_v2AWiKwnEWwQvGBG5WsC64zGNBGiyVyvGBxBan4hUAAAA](https://pyright-playground.decorator-factory.su/?gzip=H4sIAPHOr2EC_zXHPQqAMAwG0D2n-OjUgl6goKcI7oKNdOgPMUJ7eyff9kRbgc2e641celMDz56OU4kY2x_v2AWiKwnEWwQvGBG5WsC64zGNBGiyVyvGBxBan4hUAAAA)

**Reply by erictraut** [reply]
Nice! The first one (picking up the first Python code block in a github issue) is especially clever.

**Reply by erictraut** [reply]
@decorator-factory, just a heads up that I finally had time to create an official [pyright playground](https://pyright-play.net/). It has a bunch of features that you had planned to include in yours.

Thanks again for creating your pyright playground. It has served the community well for the past couple of years.

**Reply by decorator-factory** [reply]
This looks great. I'm glad I have less CSS horrors in my backlog :D

Do you mind if I redirect my playground domain to https://pyright-play.net/ ?

**Reply by erictraut** [reply]
@decorator-factory, are you looking to take down your service entirely? That will break existing permalinks, right? If you're willing to keep a minimal web service running, you could add logic that converts from your URL format to mine and then redirect with the updated URL. That would preserve all existing permalinks indefinitely.

If you prefer to take down your service completely, then it's fine to just have it redirect to my site.

**Reply by decorator-factory** [reply]
I think I'll just take it down for now, maybe I'll add some conversion later

(running it is basically free because I'm using serverless containers, but I don't have a lot of free time now)

**Comment by povilasb** [comment]
That is superb @decorator-factory .

**Comment by Zzzen** [comment]
You can try https://zzzen.github.io/pyright-ast-viewer/ . It runs purely on your browser. Built on top of monaco editor, It provides basic diagnostics.

![image](https://user-images.githubusercontent.com/6630042/197330970-6becdec7-3130-4aa3-9349-815fd5bc3337.png)

**Comment by tomlarkworthy** [comment]
Hi I am thinking of running pyright in the browser for monoco help. I explored typescript/vfs but that seems overfitted to typescript language support. Or could I use that to run pyright command line in the browser and manually pushing the output to monoco (somehow)?

**Reply by erictraut** [reply]
@tomlarkworthy, can you provide more details about what you're trying to do? What use cases do you have in mind? Are you looking to replicate the functionality of the pyright playground that decorator-factory created?

Pylance (which is built on pyright) runs in [vscode.dev](https://vscode.dev/), which is a browser-based version of VS Code. Does that meet your needs?

**Reply by tomlarkworthy** [reply]
I did managed to get it all in a web-worker in the end. No server involved (beyond asset server). I took heavy inspiration from microbit's fork https://github.com/microbit-foundation/pyright/blob/microbit/THIS_FORK.md

**Reply by simontaurus** [reply]
This sounds great.  Is your code public available?

**Comment by tomlarkworthy** [comment]
No, it's a product that allows users to type python snippets in a (Monaco)
box and we would like to add help.

I should add we are not running the python clientside, it's a control plane for a python data system elsewhere
